### PR TITLE
#164069190 users should follow each other 

### PR DESCRIPTION
**Pull request by CryceTruly**
**What does this PR do**
This PR enables authenticated users to follow and unfollow one another

**Description of tasks to be completed**
- Make users follow others
- Make users unfollow others
- Display a list of following
- Display a list of my followers

#### Feedback implemented in this PR
- Add default avatar in the navbar for users who have not added a profile picture
- Add access control to allow only unauthenticated users on the signup page

**How can this be tested??**
- `clone the repo`
- `checkout ft-users-able-follow-each-other-164069190`
- `npm install && npm start`
- `log in with your account credentials`
- `Click on the navigate icon and follow a user`

**Any background context you want to provide**
N/A

Screenshot 2019-04-25 at 11 22 27

> After the following someone

Screenshot 2019-04-25 at 11 22 06
> A user can view their floowers list
Screenshot 2019-04-25 at 11 11 59

> A user can view their following list
Screenshot 2019-04-25 at 11 11 16

**What are the relevant pivotal tracker stories**
- [#164069190 ](https://www.pivotaltracker.com/story/show/164069190 )

### CODE REVIEW

File: src/__tests__/actions/followUnfollow/followUnfollow.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import {
+  followUser, getUserfollowers, getUserfollowing, unFollowUser, 
+} from "../../../actions/discover/discover";
+
+Enzyme.configure({ adapter: new EnzymeAdapter() });
+const middleware = [thunk];
+const mockStore = configureMockStore(middleware);
+const BASE_URL = process.env.BASE_URL;
+
+
```
Comment (Peace-Apple) [inline src/__tests__/actions/followUnfollow/followUnfollow.js:18]: Great work @CryceTruly however, please remove these extra lines.

File: src/__tests__/reducers/followUnfollow/followUnfollow.test.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  it("should Load a users following list", () => {
+    const successAction = {
+      type: FOLLOWING_LIST_LOADED,
+      payload: [{
+        username: "crycetruly",
+        email: "crycetruly@gmail.com",
+        bio: "this is the bio",
+        image: "testimage",
+      }],
+
```
Comment (Peace-Apple) [inline src/__tests__/reducers/followUnfollow/followUnfollow.test.js]: This too

File: src/__tests__/reducers/followUnfollow/followUnfollow.test.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+      followers: [],
+    };
+    expect(followUnfollow(initialState, successAction)).toEqual(successState);
+  });
+
+  it("should fail to Load a users following list", () => {
+    const successAction = {
+      type: FOLLOWING_LIST_LOADED_FAILED,
+      payload: "Something went wrong while loading following list",
+
```
Comment (Peace-Apple) [inline src/__tests__/reducers/followUnfollow/followUnfollow.test.js]: This too

File: src/__tests__/reducers/followUnfollow/followUnfollow.test.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    };
+
+    const successState = {
+      followers: [],
+      isFollowingSomeone: false,
+      msg: "Something went wrong while loading following list",
+      following: [],
+      isLoading: false,
+      isFollowingSomeone: false,
+
```
Comment (Peace-Apple) [inline src/__tests__/reducers/followUnfollow/followUnfollow.test.js]: This too

File: src/__tests__/reducers/followUnfollow/followUnfollow.test.js [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+  it("should Load a users followers list", () => {
+    const successAction = {
+      type: FOLLOWERS_LOADED,
+      payload: [{
+        username: "crycetruly",
+        email: "crycetruly@gmail.com",
+        bio: "this is the bio",
+        image: "testimage",
+      }],
+
```
Comment (Peace-Apple) [inline src/__tests__/reducers/followUnfollow/followUnfollow.test.js]: This too

File: src/views/profile/followersList.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+                      <p className="mb-1 text-truncate">{item.bio}</p>
+                    </div>
+
+                  </div>
+
+                  <small>{<Moment fromNow>{item.followed_at}</Moment>}</small>
+                </div>
+
+              </div>
+
```
Comment (Peace-Apple) [inline src/views/profile/followersList.js:42]: These too

File: src/views/profile/followersList.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+            ))
+
+          ) : "No one is on the followers List"
+          }
+
```
Comment (Peace-Apple) [inline src/views/profile/followersList.js:48]: This one too

File: src/views/profile/followingList.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+                      <p className="mb-1 text-truncate">{item.bio ? item.bio : "No bio set"}</p>
+                    </div>
+
+                  </div>
+
+                  <small>{<Moment fromNow>{item.followed_at}</Moment>}</small>
+                </div>
+
+              </div>
+
```
Comment (Peace-Apple) [inline src/views/profile/followingList.js]: Here too and some other places, please work on that

File: src/views/profile/followingList.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+            ))
+
+          ) : "No one is on the following List"
+          }
+
```
Comment (Peace-Apple) [inline src/views/profile/followingList.js]: Here too

**Review by Peace-Apple — CHANGES_REQUESTED** [review]

File: src/views/profile/followersList.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
+import React from "react";
+import PropTypes from "prop-types";
+import Moment from "react-moment";
```
Comment (wasibani-roy) [inline src/views/profile/followersList.js:2]: What does this do?

File: src/views/profile/followingList.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
+import React from "react";
+import PropTypes from "prop-types";
+import Moment from "react-moment";
+import ReactTimeAgo from "react-time-ago";
```
Comment (wasibani-roy) [inline src/views/profile/followingList.js:3]: What is this and how is it used??

**Review by wasibani-roy — CHANGES_REQUESTED** [review]
Good work @CryceTruly just a few questions that need to be answered

File: public/index.html [hunk 12]
```diff
@@ -8,12 +8,12 @@
 	 <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.8.1/css/all.css" integrity="sha384-50oBUHEmvpQ+1lW4y57PTFmhCaXp0ML5d60M1M7uH2+nqUivzIebhndOJK28anvf" crossorigin="anonymous">
```
Comment (BagzieGracious) [inline public/index.html]: why did you remove font-awesome link?

**Review by BagzieGracious — CHANGES_REQUESTED** [review]

File: src/__tests__/views/articleDetailView/__test__/__snapshots__/articleDetailView.test.js.snap [hunk 13]
```diff
@@ -63,7 +63,7 @@ ShallowWrapper {
     "nodeType": "function",
     "props": Object {
       "children": Array [
-        <Connect(Navbar) />,
+        <Navbar />,
```
Comment (BagzieGracious) [inline src/__tests__/views/articleDetailView/__test__/__snapshots__/articleDetailView.test.js.snap]: why did you remove the connect from Navbar component ?

**Review by BagzieGracious — CHANGES_REQUESTED** [review]

File: src/views/profile/followersList.js (hunk 10 above)
Comment (CryceTruly) [reply src/views/profile/followersList.js:2]: The library helps to Parse, validate, manipulate, and display dates and times in JavaScript.
We use it here to achieve the time ago effect.

File: src/views/profile/followingList.js (hunk 11 above)
Comment (CryceTruly) [reply src/views/profile/followingList.js:3]: Hey @wasibani-roy, the react time ago library is used for generating localized relative time strings. When the react-time-ago package is installed the javascript-time-ago package is installed automatically along with it. This works well together with the moment package to achieve the time since feature on dates.

File: public/index.html (hunk 12 above)
Comment (CryceTruly) [reply public/index.html]: I removed the font awesome cdn link because of it was being repeated in the index.html file, so I kept one instead.

File: src/__tests__/views/articleDetailView/__test__/__snapshots__/articleDetailView.test.js.snap (hunk 13 above)
Comment (CryceTruly) [reply src/__tests__/views/articleDetailView/__test__/__snapshots__/articleDetailView.test.js.snap]: It's because the navbar does not need to connect to redux.

**Review by BagzieGracious — APPROVED** [review]

**Review by Peace-Apple — APPROVED** [review]

**Review by wasibani-roy — APPROVED** [review]

File: src/__tests__/actions/followUnfollow/followUnfollow.js [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+      },
+      type: UNFOLLOWING_USER_SUCCESS,
+    }];
+    return store.dispatch(unFollowUser("username"))
+      .then(() => expect(store.getActions())
+        .toEqual(expectedActions));
+  });
+
+  it("should load all the user`s followers", () => {
+    fetchMock.getOnce("https://ah-backend-aquaman-staging.herokuapp.com/api/profiles/test/followers/", {
```
Comment (AmosWels) [inline src/__tests__/actions/followUnfollow/followUnfollow.js:80]: consider improving the indentation here

File: src/__tests__/actions/followUnfollow/followUnfollow.js [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+       [],
+      type: FOLLOWERS_LOADED,
+    }];
+    return store.dispatch(getUserfollowers("test"))
+      .then(() => expect(store.getActions())
+        .toEqual(expectedActions));
+  });
+
+  it("should load all the user`s following list", () => {
+    fetchMock.getOnce("https://ah-backend-aquaman-staging.herokuapp.com/api/profiles/test/follows/", {
```
Comment (AmosWels) [inline src/__tests__/actions/followUnfollow/followUnfollow.js:97]: same indentation issue here

File: src/__tests__/components/profile/EditProfile.test.js [hunk 16]
```diff
@@ -1,11 +1,13 @@
...
 import { mount } from "enzyme";
-import { Provider } from "react-redux";
 import store from "../../../store";
+import { BrowserRouter as Router } from "react-router-dom";
+import { Provider } from "react-redux";
 import { EditProfile } from "../../../components/profile/editProfile";
 
 describe("edit profile component tests", () => {
+  let wrapper;
+  let wrapped;
```
Comment (AmosWels) [inline src/__tests__/components/profile/EditProfile.test.js]: consider giving better variable names. Instead of `wrapper`, can use `wrapperComponent`, or `wrapperInstance` or any other

File: src/__tests__/reducers/followUnfollow/followUnfollow.test.js [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+      followers: [],
+      isFollowingSomeone: true,
+      msg: null,
+      following: [],
+      isLoading: false,
+    };
+    expect(followUnfollow(initialState, action)).toEqual(loadedState);
+  });
+
+  it("should follow a user successfully", () => {
```
Comment (AmosWels) [inline src/__tests__/reducers/followUnfollow/followUnfollow.test.js:40]: can change this title to `should allow a user to successfully follow another user`. Also consider reviewing other `it` titles

File: src/views/profile/ProfileStats.js [hunk 18]
```diff
@@ -1,20 +1,48 @@
...
+        {followers.follower_count === 1 ? "Follower" : "Followers"}
+      </div>
+      <div className="col-md-4" data-toggle="modal" data-target="#exampleModal">
+        <strong>
+          {following.following_count ? following.following_count : "0"}
+          {" "}
+        </strong>
       Following
+      </div>
+
```
Comment (AmosWels) [inline src/views/profile/ProfileStats.js]: remove all these unnecessary lines

**Review by AmosWels — CHANGES_REQUESTED** [review]
Hey @CryceTruly , this is good work. Please make sure you resolve the eslint issues on this PR as well as implement the reviews.

Also properly format your PR. Check the typo at `> A user can view their floowers list`

File: src/__tests__/actions/followUnfollow/followUnfollow.js (hunk 14 above)
Comment (CryceTruly) [reply src/__tests__/actions/followUnfollow/followUnfollow.js:80]: Thanks, @AmosWels, I have refactored indentation here.

File: src/__tests__/actions/followUnfollow/followUnfollow.js (hunk 15 above)
Comment (CryceTruly) [reply src/__tests__/actions/followUnfollow/followUnfollow.js:97]: Thanks, @AmosWels, I have refactored indentation here.

File: src/__tests__/components/profile/EditProfile.test.js (hunk 16 above)
Comment (CryceTruly) [reply src/__tests__/components/profile/EditProfile.test.js]: Thanks @AmosWels ,have renamed these variables.

**Review by AmosWels — APPROVED** [review]
